### PR TITLE
Refactor test suite

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

**Comment by coot** [comment]
You need to rebase this PR, parts of are already merged isn't it?

File: io-sim/io-sim.cabal [hunk 1]
```diff
@@ -78,9 +78,10 @@ test-suite test
   type:                exitcode-stdio-1.0
   hs-source-dirs:      test
   main-is:             Main.hs
-  other-modules:       Test.IOSim
-                       Test.STM
-                       Test.Control.Monad.IOSimPOR
+  other-modules:       Test.STM
```
Comment (coot) [inline io-sim/io-sim.cabal:81]: Maybe:
```suggestion
  other-modules:       Test.Control.Monad.STM
```

File: io-sim/io-sim.cabal [hunk 2]
```diff
@@ -78,9 +78,10 @@ test-suite test
...
+                       Test.Control.Monad.Common
+                       Test.Control.Monad.IOSim.Test
+                       Test.Control.Monad.IOSimPOR.Test
```
Comment (coot) [inline io-sim/io-sim.cabal:84]:
```suggestion
                       Test.Control.Monad.IOSim
                       Test.Control.Monad.IOSimPOR
```

File: io-sim/test/Test/Control/Monad/Common.hs [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+-- a trace of events we want to observe, and can do probe output from
+-- multiple threads.
+--
+type Probe m x = StrictTVar m [x]
+
+withProbe :: MonadSTM m => (Probe m x -> m ()) -> m [x]
+withProbe action = do
+    probe <- newTVarIO []
+    action probe
+    reverse <$> atomically (readTVar probe)
```
Comment (coot) [inline io-sim/test/Test/Control/Monad/Common.hs:267]: we used this pattern when we didn't had `Dynamic` tracing ...

File: io-sim/test/Test/Control/Monad/Common.hs [hunk 4]
```diff
@@ -0,0 +1,509 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Control.Monad.Common where
```
Comment (coot) [inline io-sim/test/Test/Control/Monad/Common.hs:6]: :thinking:  `Test.Control.Monad.Utils` might be a better name.

File: io-sim/test/Test/Control/Monad/Common.hs (hunk 3 above)
Comment (bolt12) [reply io-sim/test/Test/Control/Monad/Common.hs]: I didn't introduce this, I just refactored the code. If this needs to change maybe it should be done in a different PR

Comment (coot) [reply io-sim/test/Test/Control/Monad/Common.hs]: I know, I am just explaining why it's there.

File: io-sim/io-sim.cabal [hunk 5]
```diff
@@ -78,9 +78,11 @@ test-suite test
   type:                exitcode-stdio-1.0
   hs-source-dirs:      test
   main-is:             Main.hs
-  other-modules:       Test.IOSim
-                       Test.STM
+  other-modules:       Test.Control.Monad.STM
+                       Test.Control.Monad.Utils
+                       Test.Control.Monad.IOSim
                        Test.Control.Monad.IOSimPOR
+                       Test.Control.Monad.IOSimPOR.Test
```
Comment (coot) [inline io-sim/io-sim.cabal:85]:
```suggestion
```

File: io-sim/test/Test/Control/Monad/IOSimPOR.hs [hunk 6]
```diff
@@ -1,40 +1,128 @@
...
+      ]
+    ]
+  ]
 
 data Step =
     WhenSet Int Int
   | ThrowTo Int
   | Delay Int
   | Timeout TimeoutStep
-  | CheckStatus Int
```
Comment (coot) [inline io-sim/test/Test/Control/Monad/IOSimPOR.hs:37]: Why it's removed in this PR when we just move files around?

Comment (bolt12) [reply io-sim/test/Test/Control/Monad/IOSimPOR.hs:37]: Good question hmmm I'll look into this

Comment (bolt12) [reply io-sim/test/Test/Control/Monad/IOSimPOR.hs]: Solved!

**Review by coot — APPROVED** [review]
LGTM

**Comment by bolt12** [comment]
bors merge

**Pull request merged**